controllers/vspheremachine_controller.go

```diff
@@ -151,35 +158,109 @@ func (r machineReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr er
 			vsphereMachine.Name)
 	}
 
-	// Create the cluster context for this request.
-	clusterContext := &context.ClusterContext{
-		ControllerContext: r.ControllerContext,
-		Cluster:           cluster,
-		VSphereCluster:    vsphereCluster,
-		Logger:            r.Logger.WithName(req.Namespace).WithName(req.Name),
-		PatchHelper:       patchHelper,
-	}
-
 	// Create the machine context for this request.
 	machineContext := &context.MachineContext{
-		ClusterContext: clusterContext,
+		ClusterContext: &context.ClusterContext{
+			ControllerContext: r.ControllerContext,
+			Cluster:           cluster,
+			VSphereCluster:    vsphereCluster,
+		},
 		Machine:        machine,
 		VSphereMachine: vsphereMachine,
 		Session:        authSession,
 		Logger:         r.Logger.WithName(req.Namespace).WithName(req.Name),
 		PatchHelper:    patchHelper,
 	}
 
+	// Print the task-ref upon entry and upon exit.
+	machineContext.Logger.V(4).Info(
+		"VSphereMachine.Status.TaskRef OnEntry",
+		"task-ref", machineContext.VSphereMachine.Status.TaskRef)
+	defer func() {
+		machineContext.Logger.V(4).Info(
+			"VSphereMachine.Status.TaskRef OnExit",
+			"task-ref", machineContext.VSphereMachine.Status.TaskRef)
+	}()
+
 	// Always issue a patch when exiting this function so changes to the
 	// resource are patched back to the API server.
 	defer func() {
+		// Patch the VSphereMachine resource.
 		if err := machineContext.Patch(); err != nil {
 			if reterr == nil {
 				reterr = err
-			} else {
-				machineContext.Logger.Error(err, "patch failed", "machine", machineContext.String())
 			}
+			machineContext.Logger.Error(err, "patch failed", "machine", machineContext.String())
 		}
+
+		// localObj is a deep copy of the VSphereMachine resource that was
+		// fetched at the top of this Reconcile function.
+		localObj := machineContext.VSphereMachine.DeepCopy()
```

Comment:
I'm not sure we need to deepcopy here. Maybe just store the vsphereMachine as you are reading status and RV

Reply:
We already have the `VSphereMachine` as it is in `machineContext.VSphereMachine`. I deep copied it during debugging. It was originally just a short-cut reference to the resource in the context.

If the current tests pass I'm going to let this merge as-is, and we can clean this up later if that's okay with you?

Reply:
yup definitely not a blocker